File: java/Figuras/src/main/java/org/example/PersistenciaArchivos.java
package org.example;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import java.io.IOException;
import java.lang.reflect.Type;
import java.nio.charset.StandardCharsets;
import java.nio.file.*;
import java.util.*;

public class PersistenciaArchivos {

    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();

    public static boolean guardarEnArchivo(Map<Integer, Figura> figuras, String archivo) {
        try {
            Path path = Paths.get(archivo);
            Path parent = path.getParent();
            if (parent != null && !Files.exists(parent)) {
                Files.createDirectories(parent);
            }
            //System.out.println("Guardando JSON en: " + path.toAbsolutePath());
            List<Map<String, Object>> datos = new ArrayList<>();
            for (Figura f : figuras.values()) {
                datos.add(figuraADict(f));
            }
            String json = gson.toJson(datos);
            Files.write(path, json.getBytes(StandardCharsets.UTF_8),
                        StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING);
            //System.out.println("JSON guardado correctamente.");
            return true;
        } catch (IOException e) {
            System.err.println("Error al guardar archivo: " + e.getMessage());
            return false;
        }
    }

    public static List<Figura> cargarDesdeArchivo(String archivo) {
        List<Figura> figuras = new ArrayList<>();
        Path path = Paths.get(archivo);
        System.out.println("Cargando JSON desde: " + path.toAbsolutePath());
        if (!Files.exists(path)) {
            System.out.println("Archivo no existe, devolviendo lista vacía.");
            return figuras;
        }
        try {
            String json = Files.readString(path, StandardCharsets.UTF_8);
            Type listType = new TypeToken<List<Map<String, Object>>>(){}.getType();
            List<Map<String, Object>> datosFiguras = gson.fromJson(json, listType);
            if (datosFiguras != null) {
                for (Map<String, Object> datos : datosFiguras) {
                    Figura f = dictAFigura(datos);
                    if (f != null) {
                        figuras.add(f);
                    }
                }
            }
            System.out.println("JSON cargado, figuras encontradas: " + figuras.size());
        } catch (IOException e) {
            System.err.println("Error al cargar archivo: " + e.getMessage());
        }
        return figuras;
    }

    // Helper: convierte Figura a mapa
    private static Map<String, Object> figuraADict(Figura figura) {
        Map<String, Object> datos = new HashMap<>();
        datos.put("id", figura.getId());
        datos.put("nombre", figura.getNombre());
        datos.put("tipo", figura.getTipo());
        if (figura instanceof Circulo) {
            datos.put("radio", ((Circulo) figura).getRadio());
        } else if (figura instanceof Cuadrado) {
            datos.put("lado", ((Cuadrado) figura).getLado());
        } else if (figura instanceof Cubo) {
            datos.put("lado", ((Cubo) figura).getLado());
        } else if (figura instanceof Esfera) {
            datos.put("radio", ((Esfera) figura).getRadio());
        }
        return datos;
    }

    // Helper: convierte mapa a Figura
    private static Figura dictAFigura(Map<String, Object> datos) {
        try {
            String nombre = datos.get("nombre").toString().toLowerCase();
            int id = ((Number) datos.get("id")).intValue();
            if (datos.containsKey("radio")) {
                double radio = ((Number) datos.get("radio")).doubleValue();
                if (nombre.equals("círculo")) {
                    return new Circulo(radio, id);
                } else if (nombre.equals("esfera")) {
                    return new Esfera(radio, id);
                }
            } else if (datos.containsKey("lado")) {
                double lado = ((Number) datos.get("lado")).doubleValue();
                if (nombre.equals("cuadrado")) {
                    return new Cuadrado(lado, id);
                } else if (nombre.equals("cubo")) {
                    return new Cubo(lado, id);
                }
            }
        } catch (Exception e) {
            System.err.println("Error al convertir datos a figura: " + e.getMessage());
        }
        return null;
    }
}

File: python/Figuras/PersistenciaArchivos.py
"""
Manejo de persistencia de datos en archivos
"""
# pylint: disable=invalid-name
import json
import os
from typing import List, Dict, Any
from Figura import Figura
from FiguraFactory import FiguraFactory

class PersistenciaArchivos:
    """Clase para manejar la persistencia de figuras en archivos JSON"""

    @staticmethod
    def guardar_en_archivo(figuras: Dict[int, Figura], archivo: str) -> bool:
        """
        Guarda las figuras en un archivo JSON

        Args:
            figuras (Dict[int, Figura]): Diccionario de figuras a guardar
            archivo (str): Ruta del archivo donde guardar

        Returns:
            bool: True si se guardó exitosamente, False en caso contrario
        """
        try:
            # Crear directorio si no existe
            directorio = os.path.dirname(archivo)
            if directorio and not os.path.exists(directorio):
                os.makedirs(directorio)

            # Convertir figuras a formato serializable
            datos_figuras = []
            for figura in figuras.values():
                datos_figura = PersistenciaArchivos._figura_a_dict(figura)
                datos_figuras.append(datos_figura)

            # Guardar en archivo JSON
            with open(archivo, 'w', encoding='utf-8') as f:
                json.dump(datos_figuras, f, indent=2, ensure_ascii=False)

            return True

        except Exception as e:
            print(f"Error al guardar archivo: {e}")
            return False

    @staticmethod
    def cargar_desde_archivo(archivo: str) -> List[Figura]:
        """
        Carga figuras desde un archivo JSON

        Args:
            archivo (str): Ruta del archivo a cargar

        Returns:
            List[Figura]: Lista de figuras cargadas
        """
        figuras = []

        try:
            if not os.path.exists(archivo):
                return figuras

            with open(archivo, 'r', encoding='utf-8') as f:
                datos_figuras = json.load(f)

            for datos in datos_figuras:
                figura = PersistenciaArchivos._dict_a_figura(datos)
                if figura:
                    figuras.append(figura)

        except Exception as e:
            print(f"Error al cargar archivo: {e}")

        return figuras

    @staticmethod
    def _figura_a_dict(figura: Figura) -> Dict[str, Any]:
        """
        Convierte una figura a diccionario para serialización

        Args:
            figura (Figura): Figura a convertir

        Returns:
            Dict[str, Any]: Diccionario con datos de la figura
        """
        datos = {
            'id': figura.get_id(),
            'nombre': figura.get_nombre(),
            'tipo': figura.get_tipo()
        }

        # Agregar atributos específicos según el tipo
        if hasattr(figura, 'get_radio'):
            datos['radio'] = figura.get_radio()
        elif hasattr(figura, 'get_lado'):
            datos['lado'] = figura.get_lado()

        return datos

    @staticmethod
    def _dict_a_figura(datos: Dict[str, Any]) -> Figura:
        """
        Convierte un diccionario a figura

        Args:
            datos (Dict[str, Any]): Diccionario con datos de la figura

        Returns:
            Figura: Figura creada o None si hay error
        """
        try:
            nombre = datos['nombre'].lower()

            if 'radio' in datos:
                if nombre == 'círculo':
                    from Circulo import Circulo
                    figura = Circulo(datos['radio'], datos['id'])
                elif nombre == 'esfera':
                    from Esfera import Esfera
                    figura = Esfera(datos['radio'], datos['id'])
                else:
                    return None
            elif 'lado' in datos:
                if nombre == 'cuadrado':
                    from Cuadrado import Cuadrado
                    figura = Cuadrado(datos['lado'], datos['id'])
                elif nombre == 'cubo':
                    from Cubo import Cubo
                    figura = Cubo(datos['lado'], datos['id'])
                else:
                    return None
            else:
                return None

            return figura

        except Exception as e:
            print(f"Error al convertir datos a figura: {e}")
            return None
